File: KantechGatewayApp/Infrastructure/Logger.vb
Imports System.IO
Namespace KantechGatewayApp.Infrastructure
    Public NotInheritable Class Logger
        Private Shared ReadOnly logRoot As String = IO.Path.GetFullPath(System.Configuration.ConfigurationManager.AppSettings("Root.Logs"))

        Private Shared Function LogPath(Optional subFolder As String = Nothing) As String
            Dim fname = If(Boolean.Parse(System.Configuration.ConfigurationManager.AppSettings("Log.RollDaily")), $"app_{Date.Today:yyyyMMdd}.log", "app.log")
            Dim logDir = If(String.IsNullOrEmpty(subFolder), logRoot, Path.Combine(logRoot, subFolder))
            Directory.CreateDirectory(logDir)
            Return Path.Combine(logDir, fname)
        End Function

        Public Shared Sub Info(msg As String, Optional subFolder As String = Nothing)
            Write("INFO", msg, subFolder)
        End Sub
        Public Shared Sub Warn(msg As String, Optional subFolder As String = Nothing)
            Write("WARN", msg, subFolder)
        End Sub
        Public Shared Sub [Error](msg As String, Optional ex As Exception = Nothing, Optional subFolder As String = Nothing)
            Write("ERROR", If(ex Is Nothing, msg, $"{msg} :: {ex}"), subFolder)
        End Sub

        Private Shared Sub Write(level As String, msg As String, subFolder As String)
            Dim line = $"{DateTime.Now:yyyy-MM-dd HH:mm:ss} [{level}] {msg}"
            SyncLock GetType(Logger)
                File.AppendAllText(LogPath(subFolder), line & Environment.NewLine)
            End SyncLock
        End Sub
    End Class
End Namespace

File: KantechGatewayApp/Infrastructure/Logger.cs
using System;
using System.IO;

namespace KantechGatewayApp.KantechGatewayApp.Infrastructure
{
    public sealed class Logger
    {
        private static readonly string logRoot = Path.GetFullPath(System.Configuration.ConfigurationManager.AppSettings["Root.Logs"]);

        private static string LogPath(string subFolder = null)
        {
            string fname = bool.Parse(System.Configuration.ConfigurationManager.AppSettings["Log.RollDaily"]) ? $"app_{DateTime.Today:ddMMyyyy}.log" : "app.log";
            string logDir = string.IsNullOrEmpty(subFolder) ? logRoot : Path.Combine(logRoot, subFolder);
            Directory.CreateDirectory(logDir);
            return Path.Combine(logDir, fname);
        }

        public static void Info(string msg, string subFolder = null)
        {
            Write("INFO", msg, subFolder);
        }
        public static void Warn(string msg, string subFolder = null)
        {
            Write("WARN", msg, subFolder);
        }
        public static void Error(string msg, Exception ex = null, string subFolder = null)
        {
            Write("ERROR", ex is null ? msg : $"{msg} :: {ex}", subFolder);
        }

        private static void Write(string level, string msg, string subFolder)
        {
            string line = $"{DateTime.Now:yyyy-MM-dd HH:mm:ss} [{level}] {msg}";
            lock (typeof(Logger))
                File.AppendAllText(LogPath(subFolder), line + Environment.NewLine);
        }
    }
}